dds/DCPS/XTypes/IdlScanner.h

```diff
@@ -0,0 +1,890 @@
+    return token;
+  }
+
+  static IdlToken make_identifier(const std::string& value)
+  {
+    IdlToken token;
+    if (value == "TRUE") {
+      token.kind_ = BooleanLiteral;
+      token.boolean_value_ = true;
+    } else if (value == "FALSE") {
+      token.kind_ = BooleanLiteral;
+      token.boolean_value_ = false;
+    } else {
+      token.kind_ = Identifier;
+      token.identifier_value_ = value;
+    }
+    return token;
+  }
+
+  bool operator==(const IdlToken& other) const
+  {
+    if (kind_ != other.kind_) {
+      return false;
+    }
+
+    switch (kind_) {
+    case Error:
+    case EOI:
+      return true;
+    case BooleanLiteral:
+      return boolean_value_ == other.boolean_value_;
+    case IntegerLiteral:
+      return signed_ == other.signed_ && integer_value_ == other.integer_value_;
+    case FloatLiteral:
+      return signed_ == other.signed_ && integer_value_ == other.integer_value_ && exponent_signed_ == other.exponent_signed_ && exponent_value_ == other.exponent_value_;
+    case StringLiteral:
+      return string_value_ == other.string_value_;
+    case CharacterLiteral:
+      return character_value_ == other.character_value_;
+    case Identifier:
+      return identifier_value_ == other.identifier_value_;
+    default:
+      return false;
+    }
+  }
+
+  bool is_error() const
+  {
+    return kind_ == Error;
+  }
+
+  bool is_boolean() const
+  {
+    return kind_ == BooleanLiteral;
+  }
+
+  bool is_integer() const
+  {
+    return kind_ == IntegerLiteral;
+  }
+
+  bool is_float() const
+  {
+    return kind_ == FloatLiteral;
+  }
+
+  bool is_identifier() const
+  {
+    return kind_ == Identifier;
+  }
+
+  bool is_signed() const
+  {
+    return signed_;
+  }
+
+  uint64_t integer_value() const
+  {
+    return integer_value_;
+  }
+
+  bool exponent_is_signed() const
+  {
+    return exponent_signed_;
+  }
+
+  uint64_t exponent_value() const
+  {
+    return exponent_value_;
+  }
+
+  const std::string& identifier_value() const
+  {
+    return identifier_value_;
+  }
+
+private:
+  Kind kind_;
+  bool boolean_value_;
+  char character_value_;
+  std::string string_value_;
+  bool signed_;  // Integer or Float is signed.
+  uint64_t integer_value_; // Integer or Float.
+  bool exponent_signed_; // Exponent is signed for Float.
+  uint64_t exponent_value_; // Exponent value for Float.
+  std::string identifier_value_;
+};
+
+class OpenDDS_Dcps_Export IdlScanner {
+public:
+  IdlScanner(CharacterScanner& scanner)
+    : scanner_(scanner)
+    , character_literal_(0)
+    , signed_(false)
+    , integer_value_(0)
+    , digits_after_decimal_(0)
+    , exponent_signed_(false)
+    , exponent_value_(0)
+  {}
+
+  IdlToken scan_token() {
+    if (scanner_.eoi()) {
+      return IdlToken::make_eoi();
+    }
+
+    const char c = scanner_.peek();
+    if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z')) {
+      return scan_identifier();
+    }
+
+    switch (c) {
+    case '\'':
```

Comment:
I know we're probably not supporting it at least for the moment, but does this handle being given a wide string or char?

Reply:
I was uninterested in wide characters since it is unclear how `default_value` would store a wide character or string.